harness/tests/core/test_profiler.py

```diff
@@ -0,0 +1,265 @@
+        write_time=0,
+    )
+    mock_psutil_disk_io.return_value = mock_start_disk_io
+
+    disk_collector = profiler._Disk()
+    mock_psutil_disk_io.assert_called_once()
+    assert mock_psutil_disk_usage.call_count == 2
+    assert len(disk_collector.metric_samples) == 0
+
+    mock_end_time = mock_start_time + 1
+    mock_time.return_value = mock_end_time
+    mock_disk_io = psutil._common.sdiskio(
+        read_count=20,
+        write_count=30,
+        read_bytes=200,
+        write_bytes=300,
+        read_time=0,
+        write_time=0,
+    )
+    mock_psutil_disk_io.return_value = mock_disk_io
+
+    mock_disk_usage_paths = {}
+    mock_disk_usage = psutil._common.sdiskusage(total=100, used=10, free=90, percent=10.0)
+    for dp in disk_collector._disk_paths:
+        mock_disk_usage_paths[dp] = mock_disk_usage
+    mock_psutil_disk_usage.return_value = mock_disk_usage
+
+    exp_thru_read = (mock_disk_io.read_bytes - mock_start_disk_io.read_bytes) / (
+        mock_end_time - mock_start_time
+    )
+    exp_thru_write = (mock_disk_io.write_bytes - mock_start_disk_io.write_bytes) / (
+        mock_end_time - mock_start_time
+    )
+    exp_iops = (mock_disk_io.read_count + mock_disk_io.write_count) - (
+        mock_start_disk_io.read_count + mock_start_disk_io.write_count
+    ) / (mock_end_time - mock_start_time)
+
+    expected_metrics = {
+        "disk_iops": exp_iops,
+        "disk_throughput_read": exp_thru_read,
+        "disk_throughput_write": exp_thru_write,
+    }
+    for p, u in mock_disk_usage_paths.items():
+        expected_metrics[p] = {"disk_util": u.percent}
+
+    disk_collector.sample_metrics()
+    assert len(disk_collector.metric_samples) == 1
+    assert disk_collector.metric_samples[0] == expected_metrics
+
+
+@mock.patch("determined.core._profiler.psutil.virtual_memory")
+def test_sample_metrics_memory(
+    mock_psutil: mock.MagicMock,
+) -> None:
+    cpu_collector = profiler._Memory()
+    assert len(cpu_collector.metric_samples) == 0
+    MockMemoryInfo = NamedTuple("MockMemoryInfo", [("available", int)])
+
+    mock_psutil.return_value = MockMemoryInfo(available=900)
+
+    expected_metrics = {
+        "memory_free": 900 / 1e9,
+    }
+
+    cpu_collector.sample_metrics()
+    assert len(cpu_collector.metric_samples) == 1
+    assert cpu_collector.metric_samples[0] == expected_metrics
+
+
+@mock.patch("determined.core._profiler.psutil.cpu_percent")
+def test_sample_metrics_cpu(
+    mock_psutil: mock.MagicMock,
+) -> None:
+    cpu_collector = profiler._CPU()
+    assert len(cpu_collector.metric_samples) == 0
+
+    mock_cpu_util = 20.0
+    mock_psutil.return_value = mock_cpu_util
+
+    expected_metrics = {
+        "cpu_util_simple": mock_cpu_util,
+    }
+
+    cpu_collector.sample_metrics()
+    assert len(cpu_collector.metric_samples) == 1
+    assert cpu_collector.metric_samples[0] == expected_metrics
+
+
+@pytest.mark.skipif(profiler.pynvml is None, reason="pynvml required for test")
```

Comment:
How do we keep this tested in CI?

Reply:
this is tested by `test-unit-harness-gpu`